### PR TITLE
sched/backtrace: add support for printing the arbitrary thread backtrace in runtime

### PR DESCRIPTION
**Pull request by anchao**
## Summary

add support for printing the arbitrary thread backtrace in runtime to monitor the thread status
( only support cortex-m / cortex-a(thumb) )

libs/libc: move the backtrace implement to sched
arch/cortex-m: add arch backtrace support 
sched/backtrace: add sched_backtrace support 

## Impact

Depends on:
https://github.com/apache/incubator-nuttx-apps/pull/817

N/A, new feature

## Testing

Running tasks:

```
ap> ps
  PID GROUP PRI POLICY   TYPE    NPX STATE    EVENT     SIGMASK   STACK   USED  FILLED    CPU COMMAND
    0     0   0 FIFO     Kthread N-- Ready              00000000 003052 001044  34.2%   94.4% Idle Task
    1     1 253 RR       Kthread --- Waiting  Semaphore 00000000 003044 000344  11.3%    0.0% hpwork 0x2011dc34
    2     2 253 RR       Kthread --- Waiting  Semaphore 00000000 003044 000344  11.3%    0.0% hpwork 0x2011dc34
    3     3 253 RR       Kthread --- Waiting  Semaphore 00000000 003044 000344  11.3%    0.0% hpwork 0x2011dc34
    4     4 110 RR       Kthread --- Waiting  Semaphore 00000000 008164 000824  10.0%    1.6% lpwork 0x2011dc28
    5     5 110 RR       Kthread --- Waiting  Semaphore 00000000 008164 001592  19.5%    0.0% lpwork 0x2011dc28
    6     6 110 RR       Kthread --- Waiting  Semaphore 00000000 008164 001544  18.9%    0.5% lpwork 0x2011dc28
    8     8 101 RR       Kthread --- Waiting  Semaphore 00000000 004060 001896  46.6%    0.0% bes_main  0x34008650
    9     9 252 RR       Kthread --- Waiting  MQ empty  00000000 006108 000488   7.9%    0.0% app_thread  0x34007924
   10    10 253 RR       Kthread --- Waiting  Semaphore 00000000 008124 000504   6.2%    0.0% audio_flinger  0x20042968
   11    11 100 RR       Kthread --- Waiting  Signal    00000000 003028 000472  15.5%    0.0% apps_recover  0x34007944
   12    12 252 RR       Kthread --- Waiting  MQ empty  00000000 010180 000504   4.9%    0.0% net_wq  0x20044d60
   13    13 252 RR       Kthread --- Waiting  MQ empty  00000000 008132 000480   5.9%    0.0% net_tasklet  0x20044d28
   14    14 252 RR       Kthread --- Waiting  Semaphore 00000000 004036 000920  22.7%    0.0% cw1200_bh  0x20044e14
   15    15 224 RR       Kthread --- Waiting  Semaphore 00000000 004060 000636  15.6%    0.0% rptun audio 0x20147250
   16    16 100 RR       Task    --- Waiting  Semaphore 00000000 004060 000368   9.0%    0.0% rc_raw_event_thx 0x20148f70
   17    17 100 RR       Task    --- Running            00000000 004084 001824  44.6%    0.5% init
   18    18 100 RR       Task    --- Waiting  Semaphore 00000000 004084 002384  58.3%    0.0% kvdbd
   20    20 252 RR       Task    --- Waiting  Semaphore 00000000 004060 000536  13.2%    0.0% rpmsg-uorb-audio 0x20157650
   21    21 232 RR       Task    --- Waiting  Semaphore 00000000 006132 001808  29.4%    0.0% usrsock
   23    23 227 FIFO     Kthread --- Waiting  Semaphore 00000000 001984 000712  35.8%    0.0% BT TX  0x201575f0
   24    24 228 FIFO     Kthread --- Waiting  Semaphore 00000000 004032 000880  21.8%    0.5% BT RX  0x201575f0
   25    25 230 FIFO     Kthread --- Waiting  Semaphore 00000000 001984 000576  29.0%    0.0% BT ECC  0x201575f0
   26    26 228 FIFO     Kthread --- Waiting  Semaphore 00000000 004032 000944  23.4%    2.7% BT Driver  0x201575f0
   27    27 101 RR       Kthread --- Waiting  MQ empty  00000000 004060 000608  14.9%    0.0% wifi_event  0x34008524
   28    28 252 RR       Kthread --- Waiting  MQ empty  00000000 004052 000568  14.0%    0.0% bes_netdev_wq
   29    29 100 RR       Kthread --- Waiting  Signal    00000000 004052 000520  12.8%    0.0% trans_stat_task  0x34008504
   30    30 100 RR       Kthread --- Ready              00000000 003036 000552  18.1%    0.0% temp_main  0x34008630

```

show task 1 hpwork:

```
ap> dumpstack 1
[06:14:18] [32] [BackTrace| 1|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
```

check the backtrace:
```
arm-none-eabi-addr2line -e nuttx 0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
/home/archer/code/m3/nuttx/arch/arm/src/armv8-m/arm_switchcontext.S:79
/home/archer/code/m3/nuttx/sched/semaphore/sem_wait.c:153 (discriminator 2)
/home/archer/code/m3/nuttx/sched/semaphore/sem_wait.c:222 (discriminator 1)
/home/archer/code/m3/nuttx/sched/wqueue/kwork_thread.c:132
/home/archer/code/m3/nuttx/sched/task/task_start.c:139
```

dump stack in range 0 - 30:

```
ap> dumpstack 0 30
[06:14:27] [33] [BackTrace| 0|0]:    0x240420   0x240184 0x2400460b
[06:14:27] [33] [BackTrace| 1|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 2|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 3|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 4|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 5|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 6|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0d7b06   0x2128a8
[06:14:27] [33] [BackTrace| 8|0]:   0xc0e635c   0x2111b6   0x2111d6   0x2103a4   0x210170   0x222556  0xc1ddad0   0x2128a8
[06:14:27] [33] [BackTrace| 9|0]:   0xc0e635c   0x20f8b6   0x20f68e   0x222b1e  0xc1b066a  0xc1b0696   0x2128a8
[06:14:27] [33] [BackTrace|10|0]:   0xc0e635c   0x2111b6   0x2111d6   0x2103a4   0x210170   0x222556   0x20df16   0x2128a8
[06:14:27] [33] [BackTrace|11|0]:   0xc0e635c   0x211b6a   0x2121a8   0x212268  0xc0e5080  0xc1f6724   0x2128a8
[06:14:27] [33] [BackTrace|12|0]:   0xc0e635c   0x20f8b6   0x20f68e   0x222b1e   0x22d968   0x2128a8
[06:14:27] [33] [BackTrace|13|0]:   0xc0e635c   0x20f8b6   0x20f68e   0x2229ac   0x22e702   0x2128a8
[06:14:27] [33] [BackTrace|14|0]:   0xc0e635c   0x2111b6   0x2111e2   0x222732   0x23e90e   0x2128a8
[06:14:27] [33] [BackTrace|15|0]:   0xc0e635c   0x2111b6   0x2113a0   0x2113da  0xc0daa9c   0x2128a8
[06:14:27] [33] [BackTrace|16|0]:   0xc0e635c   0x2111b6  0xc0e7b62  0xc0e07e4   0x2128bc
[06:14:27] [33] [BackTrace|17|0]:   0xc0e635c   0x211bac   0x210d38   0x210d8e  0xc0fd872  0xc0fa42a  0xc0fd4e4  0xc0fe43c
[06:14:27] [33] [BackTrace|17|1]:   0xc0fe912  0xc0e07e4   0x2128bc
[06:14:27] [33] [BackTrace|18|0]:   0xc0e635c   0x2111b6  0xc196746  0xc1967b6  0xc134324  0xc1345dc  0xc0e07e4   0x2128bc
[06:14:27] [33] [BackTrace|20|0]:   0xc0e635c   0x2111b6  0xc196746  0xc1967b6   0x2122fa  0xc108db4  0xc0e07e4   0x2128bc
[06:14:27] [33] [BackTrace|21|0]:   0xc0e635c   0x2111b6  0xc196746  0xc1967b6  0xc103e1c 0xdeadbeec
[06:14:27] [33] [BackTrace|23|0]:   0xc0e635c   0x2111b6   0x2111e2  0xc16f02e  0xc16ecee  0xc154a90  0xc16f36a   0x2128a8
[06:14:27] [33] [BackTrace|24|0]:   0xc0e635c   0x2111b6   0x2111e2  0xc16f02e  0xc16ecee  0xc16f2c2  0xc16f5f0  0xc15494c
[06:14:27] [33] [BackTrace|24|1]:   0xc16f36a   0x2128a8
[06:14:27] [33] [BackTrace|25|0]:   0xc0e635c   0x2111b6   0x2111e2  0xc16f02e  0xc15381c  0xc16f36a   0x2128a8
[06:14:27] [33] [BackTrace|26|0]:   0xc0e635c   0x2111b6   0x2111d6  0xc0e8d9e  0xc0dd790  0xc0dd82c  0xc16fb80  0xc16fbbe
[06:14:27] [33] [BackTrace|26|1]:   0xc16f36a   0x2128a8
[06:14:27] [33] [BackTrace|27|0]:   0xc0e635c   0x20f8b6   0x20f68e   0x222b1e  0xc1d4b3c   0x2128a8
[06:14:27] [33] [BackTrace|28|0]:   0xc0e635c   0x20f8b6   0x20f68e   0x223b04   0x2128a8
[06:14:27] [33] [BackTrace|29|0]:   0xc0e635c   0x211b6a   0x2121a8   0x212268  0xc0e5080  0xc1d419a   0x2128a8
[06:14:27] [33] [BackTrace|30|0]:   0xc0e635c   0x211b6a   0x2121a8   0x212268  0xc0e5080  0xc1dd748   0x2128a8
```

### CODE REVIEW

**Comment by xiaoxiang781216** [comment]
Look good, please create a new patch to incoperate xtensa bracktrace to the new interface.

File: arch/arm/src/armv6-m/arm_backtrace.c [hunk 1]
```diff
@@ -0,0 +1,266 @@
...
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ ****************************************************************************/
```
Comment (acassis) [inline arch/arm/src/armv6-m/arm_backtrace.c:72]: ditto

File: arch/arm/src/armv6-m/arm_backtrace.c [hunk 2]
```diff
@@ -0,0 +1,266 @@
...
+    {
+      return 0;
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ ****************************************************************************/
```
Comment (acassis) [inline arch/arm/src/armv6-m/arm_backtrace.c:88]: ditto

File: arch/arm/src/armv6-m/arm_backtrace.c [hunk 3]
```diff
@@ -0,0 +1,266 @@
...
+  return *ppc;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_backtrace
+ ****************************************************************************/
```
Comment (acassis) [inline arch/arm/src/armv6-m/arm_backtrace.c:204]: ditto

File: arch/arm/src/armv7-a/arm_backtrace.c [hunk 4]
```diff
@@ -0,0 +1,268 @@
...
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ ****************************************************************************/
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_backtrace.c:72]: ditto

File: arch/arm/src/armv7-a/arm_backtrace.c [hunk 5]
```diff
@@ -0,0 +1,268 @@
...
+  if (((uintptr_t)lr & 0xffffffe0) == 0xffffffe0)
+    {
+      return 0;
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_backtrace.c:87]: ditto

File: arch/arm/src/armv7-a/arm_backtrace.c [hunk 6]
```diff
@@ -0,0 +1,268 @@
...
+
+  return *ppc;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_backtrace
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_backtrace.c:203]: ditto

File: arch/arm/src/armv7-m/arm_backtrace.c [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+#define IOP_T_SUB_SP_32   0xf1ad0d00
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
```
Comment (acassis) [inline arch/arm/src/armv7-m/arm_backtrace.c:71]: ditto

File: arch/arm/src/armv7-m/arm_backtrace.c [hunk 8]
```diff
@@ -0,0 +1,266 @@
...
+  if (((uintptr_t)lr & 0xffffffe0) == 0xffffffe0)
+    {
+      return 0;
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
```
Comment (acassis) [inline arch/arm/src/armv7-m/arm_backtrace.c:87]: ditto

File: arch/arm/src/armv7-m/arm_backtrace.c [hunk 9]
```diff
@@ -0,0 +1,266 @@
...
+
+  return *ppc;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_backtrace
```
Comment (acassis) [inline arch/arm/src/armv7-m/arm_backtrace.c:203]: ditto

File: arch/arm/src/armv7-r/arm_backtrace.c [hunk 10]
```diff
@@ -0,0 +1,268 @@
...
+#define IOP_T_SUB_SP_32   0xf1ad0d00
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_backtrace.c:71]: ditto

File: arch/arm/src/armv7-r/arm_backtrace.c [hunk 11]
```diff
@@ -0,0 +1,268 @@
...
+  if (((uintptr_t)lr & 0xffffffe0) == 0xffffffe0)
+    {
+      return 0;
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_backtrace.c:87]: ditto

File: arch/arm/src/armv7-r/arm_backtrace.c [hunk 12]
```diff
@@ -0,0 +1,268 @@
...
+
+  return *ppc;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_backtrace
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_backtrace.c:203]: ditto

File: arch/arm/src/armv8-m/arm_backtrace.c [hunk 13]
```diff
@@ -0,0 +1,266 @@
...
+ ****************************************************************************/
+
+#include <nuttx/config.h>
+
+#include <nuttx/arch.h>
+
+#include "sched/sched.h"
+
+/****************************************************************************
+ * Pre-processor Definitions
```
Comment (acassis) [inline arch/arm/src/armv8-m/arm_backtrace.c:32]: ditto

File: arch/arm/src/armv8-m/arm_backtrace.c [hunk 14]
```diff
@@ -0,0 +1,266 @@
...
+#define IOP_T_SUB_SP_32   0xf1ad0d00
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
```
Comment (acassis) [inline arch/arm/src/armv8-m/arm_backtrace.c:71]: ditto

File: arch/arm/src/armv8-m/arm_backtrace.c [hunk 15]
```diff
@@ -0,0 +1,266 @@
...
+
+  return *ppc;
+}
+
+/****************************************************************************
+ * Public Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: up_backtrace
```
Comment (acassis) [inline arch/arm/src/armv8-m/arm_backtrace.c:203]: ditto

File: include/nuttx/arch.h [hunk 16]
```diff
@@ -493,6 +493,18 @@ void up_exit() noreturn_function;
 
 void up_assert(FAR const char *filename, int linenum);
 
+/****************************************************************************
+ * Name: up_backtrace
+ *
+ * Description:
+ *  Dump thread backtrace from specified tcb
+ *
```
Comment (acassis) [inline include/nuttx/arch.h:501]: please include params info and return

File: arch/arm/src/arm/arm_backtrace.c [hunk 17]
```diff
@@ -0,0 +1,268 @@
...
+#define IOP_T_SUB_SP_32   0xf1ad0d00
+
+#define INSTR_LIMIT       0x2000
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
```
Comment (acassis) [inline arch/arm/src/arm/arm_backtrace.c:71]: Please include a complete Description of the function, including params and return

**Comment by anchao** [comment]
@acassis please review again, thank you!

File: arch/arm/src/arm/arm_backtrace.c [hunk 18]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/arm/arm_backtrace.c:100]: currect => current

File: arch/arm/src/armv6-m/arm_backtrace.c [hunk 19]
```diff
@@ -0,0 +1,313 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/armv6-m/arm_backtrace.c:74]: currect => current

File: arch/arm/src/armv6-m/arm_backtrace.c [hunk 20]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/armv6-m/arm_backtrace.c:100]: ditto

File: arch/arm/src/armv7-a/arm_backtrace.c [hunk 21]
```diff
@@ -0,0 +1,315 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_backtrace.c:74]: ditto

File: arch/arm/src/armv7-a/arm_backtrace.c [hunk 22]
```diff
@@ -0,0 +1,315 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/armv7-a/arm_backtrace.c:100]: ditto

File: arch/arm/src/armv7-m/arm_backtrace.c [hunk 23]
```diff
@@ -0,0 +1,313 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/armv7-m/arm_backtrace.c:74]: ditto

File: arch/arm/src/armv7-m/arm_backtrace.c [hunk 24]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/armv7-m/arm_backtrace.c:100]: ditto

File: arch/arm/src/armv7-r/arm_backtrace.c [hunk 25]
```diff
@@ -0,0 +1,316 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_backtrace.c:74]: ditto

File: arch/arm/src/armv7-r/arm_backtrace.c [hunk 26]
```diff
@@ -0,0 +1,316 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/armv7-r/arm_backtrace.c:100]: ditto

File: arch/arm/src/armv8-m/arm_backtrace.c [hunk 27]
```diff
@@ -0,0 +1,313 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/armv8-m/arm_backtrace.c:74]: ditto

File: arch/arm/src/armv8-m/arm_backtrace.c [hunk 28]
```diff
@@ -0,0 +1,313 @@
...
+    }
+
+  return (*(FAR uint16_t *)(lr - 4) & 0xf000) == 0xf000 ? 5 : 3;
+}
+
+/****************************************************************************
+ * Name: backtrace_internal
+ *
+ * Description:
+ *  backtrace_internal()  returns the currect link address from input program
```
Comment (acassis) [inline arch/arm/src/armv8-m/arm_backtrace.c:100]: ditto

File: arch/arm/src/arm/arm_backtrace.c [hunk 29]
```diff
@@ -0,0 +1,315 @@
...
+
+/****************************************************************************
+ * Private Functions
+ ****************************************************************************/
+
+/****************************************************************************
+ * Name: getlroffset
+ *
+ * Description:
+ *  getlroffset()  returns the currect link address offset.
```
Comment (acassis) [inline arch/arm/src/arm/arm_backtrace.c:74]: ditto

**Review by xiaoxiang781216 — APPROVED** [review]

**Comment by xiaoxiang781216** [comment]
@acassis could review again?

**Comment by anchao** [comment]
let us close this PR since I have split the commit to other PR, please refer: 
https://github.com/apache/incubator-nuttx/pull/4409 
https://github.com/apache/incubator-nuttx/pull/4405

**Pull request closed without merging**